Code/GraphMol/RGroupDecomposition/RGroupCore.h

```diff
@@ -36,19 +39,84 @@ struct RCore {
-  std::unique_ptr<ROMol> replaceCoreDummiesWithMolMatches(
-      const ROMol &mol, const MatchVectType &match) const {
-    std::unique_ptr<ROMol> coreReplacedDummies(new ROMol(*core));
+  ROMOL_SPTR replaceCoreDummiesWithMolMatches(
+      bool &hasDummies, const ROMol &mol, const MatchVectType &match) const {
+    auto coreReplacedDummies = boost::make_shared<RWMol>(*core);
+    hasDummies = false;
     for (const auto &p : match) {
-      auto a = coreReplacedDummies->getAtomWithIdx(p.first);
-      if (!a->getAtomicNum() && !a->hasProp(RLABEL)) {
-        a->setAtomicNum(mol.getAtomWithIdx(p.second)->getAtomicNum());
+      auto atom = coreReplacedDummies->getAtomWithIdx(p.first);
+      if (isAnyAtomWithMultipleNeighborsOrNotUserRLabel(*atom)) {
+        hasDummies = true;
+        auto molAtom = mol.getAtomWithIdx(p.second);
+        replaceDummyAtom(*coreReplacedDummies, *atom, *molAtom);
+      }
+    }
+
+    for (auto bond : coreReplacedDummies->bonds()) {
+      if (bond->hasQuery()) {
+        hasDummies = true;
+        const auto molBond =
+            mol.getBondBetweenAtoms(match[bond->getBeginAtomIdx()].second,
+                                    match[bond->getEndAtomIdx()].second);
+        Bond newBond(molBond->getBondType());
+        coreReplacedDummies->replaceBond(bond->getIdx(), &newBond, true);
       }
     }
+
+#ifdef VERBOSE
+    std::cerr << "Original core smarts  " << MolToSmarts(*core) << std::endl;
+    std::cerr << "Dummy replaced core smarts  "
+              << MolToSmarts(*coreReplacedDummies) << std::endl;
+#endif
+
     return coreReplacedDummies;
   }
+
+  void replaceDummyAtom(RWMol &mol, Atom &atom, const Atom &other) const {
+    PRECONDITION(atom.getAtomicNum() == 0, "Atom must be dummy");
+    auto atomicNumber = other.getAtomicNum();
+    if (atom.hasQuery()) {
+      Atom newAtom(atomicNumber);
+      newAtom.setIsAromatic(other.getIsAromatic());
+      mol.replaceAtom(atom.getIdx(), &newAtom, false, true);
+    } else {
+      atom.setAtomicNum(other.getAtomicNum());
+      atom.setIsAromatic(other.getIsAromatic());
+    }
+  }
+
+  // Final core returned to user with dummy atoms and bonds set to those in the
+  // match
+  ROMOL_SPTR coreWithMatches(const ROMol &coreReplacedDummies) {
+    auto finalCore = boost::make_shared<RWMol>(*labelledCore);
+    for (size_t atomIdx = 0; atomIdx < coreReplacedDummies.getNumAtoms();
+         ++atomIdx) {
+      auto coreAtom = finalCore->getAtomWithIdx(atomIdx);
+      auto templateAtom = coreReplacedDummies.getAtomWithIdx(atomIdx);
+      auto unlabelledCoreAtom = core->getAtomWithIdx(atomIdx);
+      if (coreAtom->getAtomicNum() == 0 && templateAtom->getAtomicNum() > 0 &&
+          isAnyAtomWithMultipleNeighborsOrNotUserRLabel(*unlabelledCoreAtom)) {
+        replaceDummyAtom(*finalCore, *coreAtom, *templateAtom);
+        finalCore->getAtomWithIdx(atomIdx)->setNoImplicit(true);
+      }
+    }
+
```

Comment:
@greglandrum - Setting the bonds here works, but the resulting structure wont Kekulize.